src/replication.c

```diff
@@ -775,31 +1048,101 @@ get_non_quorum_replica_count(spec_t *spec)
 static bool
 can_replica_connect(spec_t *spec, replica_t *replica)
 {
-	int replica_count, non_quorum_count;
+	int replica_count, non_quorum_count, trusty_replica_count;
 	bool ret = true;
 
 	ASSERT(MTX_LOCKED(&spec->rq_mtx));
-
+	non_quorum_count = get_non_quorum_replica_count(spec);
+
+	//TODO: Change non_quorum variables to unknown_quorum
+	//TODO: If DRF unknown replicas are already connected to the target.
+	//      If known replica came to connect we need to remove connected
+	//      replicas(in the order non-quorum, not-in-the-known-list)
+
+	/* We need to allow DRF replicas to connect to target there might
+	 * be case where rebuilding is done at replica and marked quorum=on,
+	 * immediately after that target or replica is restarted,
+	 * in this scenario we need to allow replica to connect and update
+	 * in non-quorum list and another use case in replica scale up scenarios.
+	 */
 	replica_count = spec->healthy_rcount + spec->degraded_rcount;
-	if (replica_count >= spec->replication_factor) {
+	if ((replica_count >= spec->desired_replication_factor) ||
+	    ((replica_count + non_quorum_count) >= spec->desired_replication_factor)) {
 		REPLICA_ERRLOG("removing replica(%s:%d) since healthy: %d"
-		    " degraded: %d connected >= (%d)", replica->ip,
-		    replica->port, spec->healthy_rcount,
-		    spec->degraded_rcount, spec->replication_factor);
-		ret = false;
-	} else {
-		non_quorum_count = get_non_quorum_replica_count(spec);
-		if ((non_quorum_count + replica_count) >= MAXREPLICA) {
-			REPLICA_ERRLOG("erroring replica(%s:%d) since "
-			    "non_quorum: %d healthy: %d degraded: %d are"
-			    " connected >= %d", replica->ip, replica->port,
-			    non_quorum_count, spec->healthy_rcount,
-			    spec->degraded_rcount, MAXREPLICA);
-			ret = false;
-		}
+		    " degraded: %d non_quorum: %d connected >= (%d)\n",
+		    replica->ip, replica->port, spec->healthy_rcount,
+		    spec->degraded_rcount, non_quorum_count,
+		    spec->desired_replication_factor);
+		return false;
 	}
 
-	return ret;
+	/* known and trusty replica */
+ 	ret = check_is_it_trusty_replica(spec, replica);
+	if (ret == true)
+		return true;
+
+	/* replica replacement case */
+	ret = is_trusted_replicas_contain_replicaid(spec, replica->replica_id);
```

Comment:
This is not needed. `check_is_it_trusty_replica` already verifies by `replica_id`.

Reply:
`check_is_it_trusty_replica` compares both zvol_guid and replica_id from spec->rq list and
`is_trusted_replicas_contain_replicaid` compares only replicaId which will be helpful in replica replacement/movement scenarios.